AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Core/DataStoreList.swift

```diff
@@ -0,0 +1,172 @@
+        loadIfNeeded()
+        return elements.makeIterator()
+    }
+
+    // MARK: - Persistent Operations
+
+    @available(*, deprecated, message: "Not supported.")
+    public override func limit(_ limit: Int) -> Self {
+        // TODO handle query with limit
+        self.limit = limit
+        state = .pending
+        return self
+    }
+
+    // MARK: - Codable
+
+    required convenience init(from decoder: Decoder) throws {
+        let json = try JSONValue(from: decoder)
+
+        switch json {
+        case .array:
+            let elements = try Elements(from: decoder)
+            self.init(elements)
+        case .object(let list):
+            if case let .string(associatedId) = list["associatedId"],
+               case let .string(associatedField) = list["associatedField"] {
+                let field = Element.schema.field(withName: associatedField)
+                // TODO handle eager loaded associations with elements
+                self.init([], associatedId: associatedId, associatedField: field)
+            } else {
+                self.init(Elements())
+            }
+        default:
+            self.init(Elements())
+        }
+    }
+
+    // MARK: ModelListDecoder
+
+    public static func shouldDecode(decoder: Decoder) -> Bool {
+        let json = try? JSONValue(from: decoder)
+        if case let .object(list) = json,
+           case .string = list["associatedId"],
+           case .string = list["associatedField"] {
+            return true
+        }
+
+        if case .array = json {
+            return true
+        }
+
+        return false
+    }
+
+    public static func decode<ModelType: Model>(decoder: Decoder, modelType: ModelType.Type) -> List<ModelType> {
+        do {
+            return try DataStoreList<ModelType>.init(from: decoder)
+        } catch {
+            return List([ModelType]())
+        }
+    }
+
+    // MARK: - Asynchronous API
+
+    /// Trigger `DataStore` query to initialize the collection. This function always
+    /// fetches data from the `DataStore.query`.
+    ///
+    /// - seealso: `load()`
+    public override func load(_ completion: DataStoreCallback<Elements>) {
+        lazyLoad(completion)
+    }
+
+    // MARK: - Synchronous API
+
+    /// Trigger `DataStore` query to initialize the collection. This function always
+    /// fetches data from the `DataStore.query`. However, consumers must be aware of
+    /// the internal behavior which relies on `DispatchSemaphore` and will block the
+    /// current `DispatchQueue` until data is ready. When operating on large result
+    /// sets, prefer using the asynchronous `load(completion:)` instead.
+    ///
+    /// - Returns: the current instance after data was loaded.
+    /// - seealso: `load(completion:)`
+    public override func load() -> Self {
+        lazyLoad()
+        return self
+    }
+
+    // MARK: Combine
+
+    /// Lazy load the collection and expose the loaded `Elements` as a Combine `Publisher`.
```

Comment:
We should be clear about this--are we lazily loading as soon as this is called, or only when it receives the first subscription? Combine supports both kinds of operations, so we need to specify

Reply:
Updated comment on this, i tested a sample example in playgrounds and obvserved that lazy loading happens as soon as it's called and the publisher is then returned. The `sink` returns the result of the call that occurred already.